File: lilac/src/main/java/com/hartwig/hmftools/lilac/hla/HlaCopyNumber.java
package com.hartwig.hmftools.lilac.hla;

import static com.hartwig.hmftools.lilac.LilacConfig.LL_LOGGER;
import static com.hartwig.hmftools.lilac.LilacConstants.HLA_GENES;

import com.google.common.collect.Lists;
import com.hartwig.hmftools.common.purple.gene.GeneCopyNumber;
import com.hartwig.hmftools.common.purple.gene.GeneCopyNumberFile;
import com.hartwig.hmftools.lilac.coverage.HlaAlleleCoverage;
import com.hartwig.hmftools.lilac.coverage.HlaComplexCoverage;

import java.io.IOException;
import java.util.List;
import java.util.stream.Collectors;

public class HlaCopyNumber
{
    public final HlaAllele Allele;
    public final double CopyNumber;

    public HlaCopyNumber(final HlaAllele allele, double copyNumber)
    {
        Allele = allele;
        CopyNumber = copyNumber;
    }

    public static List<HlaCopyNumber> alleleCopyNumber(final List<HlaAllele> winners)
    {
        return winners.stream().map(x -> new HlaCopyNumber(x, 0)).collect(Collectors.toList());
    }

    public static List<HlaCopyNumber> alleleCopyNumber(
            final List<HlaAllele> winners, final String geneCopyNumberFile, final HlaComplexCoverage tumorCoverage)
    {
        if(geneCopyNumberFile.isEmpty() || tumorCoverage.getAlleleCoverage().isEmpty())
            return winners.stream().map(x -> new HlaCopyNumber(x, 0.0)).collect(Collectors.toList());

        try
        {
            List<GeneCopyNumber> geneCopyNumberList = GeneCopyNumberFile.read(geneCopyNumberFile);

            List<HlaAlleleCoverage> alleleCoverage = tumorCoverage.getAlleleCoverage();

            List<GeneCopyNumber> hlaGeneCopyNumbers = geneCopyNumberList.stream()
                    .filter(x -> HLA_GENES.contains(x.gene())).collect(Collectors.toList());

            List<HlaCopyNumber> hlaCopyNumbers = Lists.newArrayList();

            for(String gene : HLA_GENES)
            {
                String geneId = gene.substring(gene.length() - 1);
                hlaCopyNumbers.addAll(alleleCopyNumber(
                        geneCopyNumberList.stream().filter(x -> x.gene().equals(gene)).findFirst().orElse(null),
                        alleleCoverage.stream().filter(x -> x.Allele.Gene.equals(geneId)).collect(Collectors.toList())));
            }

            return hlaCopyNumbers;
        }
        catch(IOException e)
        {
            LL_LOGGER.error("failed to read gene copy number file({}): {}", geneCopyNumberFile, e.toString());
            return winners.stream().map(x -> new HlaCopyNumber(x, 0.0)).collect(Collectors.toList());
        }
    }

    private static List<HlaCopyNumber> alleleCopyNumber(final GeneCopyNumber geneCopyNumber, final List<HlaAlleleCoverage> alleleCoverage)
    {
        if(geneCopyNumber == null || alleleCoverage.size() != 2)
            return Lists.newArrayList();

        double minor = geneCopyNumber.minMinorAlleleCopyNumber();
        double major = geneCopyNumber.minCopyNumber() - minor;

        HlaAllele majorAllele = alleleCoverage.get(0).TotalCoverage >= alleleCoverage.get(1).TotalCoverage ?
                alleleCoverage.get(0).Allele : alleleCoverage.get(1).Allele;

        HlaAllele minorAllele = alleleCoverage.get(0).Allele == majorAllele ?
                alleleCoverage.get(1).Allele : alleleCoverage.get(0).Allele;

        return Lists.newArrayList(new HlaCopyNumber(majorAllele, major), new HlaCopyNumber(minorAllele, minor));
    }

}

File: lilac/src/main/kotlin/com/hartwig/hmftools/lilackt/cna/HlaCopyNumber.kt
package com.hartwig.hmftools.lilackt.cna

import com.hartwig.hmftools.common.purple.gene.GeneCopyNumber
import com.hartwig.hmftools.common.purple.gene.GeneCopyNumberFile
import com.hartwig.hmftools.lilac.coverage.HlaAlleleCoverage
import com.hartwig.hmftools.lilac.coverage.HlaComplexCoverage
import com.hartwig.hmftools.lilackt.hla.HlaAllele

data class HlaCopyNumber(val allele: HlaAllele, val copyNumber: Double) {

    companion object {
        private val genes = listOf("HLA-A", "HLA-B", "HLA-C")

        fun alleleCopyNumber(winners: List<HlaAllele>): List<HlaCopyNumber> {
            return winners.map { HlaCopyNumber(it, 0.0) }
        }

        fun alleleCopyNumber(winners: List<HlaAllele>, geneCopyNumberFile: String, tumorCoverage: HlaComplexCoverage): List<HlaCopyNumber> {
            if (geneCopyNumberFile.isEmpty() || tumorCoverage.alleleCoverage.isEmpty()) {
                return winners.map { HlaCopyNumber(it, 0.0) }
            }

            val coverage = tumorCoverage.alleleCoverage
            require(coverage.size == 6)
            val geneCopyNumbers = GeneCopyNumberFile.read(geneCopyNumberFile).filter { it.gene() in genes}.sortedBy { it.gene() }
            val aCopyNumber = alleleCopyNumber(geneCopyNumbers[0], coverage.filter { it.allele.gene == "A" })
            val bCopyNumber = alleleCopyNumber(geneCopyNumbers[1], coverage.filter { it.allele.gene == "B" })
            val cCopyNumber = alleleCopyNumber(geneCopyNumbers[2], coverage.filter { it.allele.gene == "C" })

            return (aCopyNumber + bCopyNumber + cCopyNumber).sortedBy { it.allele }
        }

        private fun alleleCopyNumber(geneCopyNumber: GeneCopyNumber, alleleCoverage: List<HlaAlleleCoverage>): List<HlaCopyNumber> {
            require(alleleCoverage.size == 2)

            val minor = geneCopyNumber.minMinorAlleleCopyNumber()
            val major = geneCopyNumber.minCopyNumber() - minor

            return if (alleleCoverage[0].totalCoverage >= alleleCoverage[1].totalCoverage) {
               listOf(HlaCopyNumber(alleleCoverage[0].allele, major), HlaCopyNumber(alleleCoverage[1].allele, minor))
            } else {
                listOf(HlaCopyNumber(alleleCoverage[0].allele, minor), HlaCopyNumber(alleleCoverage[1].allele, major))
            }
        }

    }

}
